### PR TITLE
Bugfix/58/h3 parent approach

### LINKED ISSUE #58: Raise exception when no data is returned for small features
As a workaround to https://github.com/worldbank/DECAT_Space2Stats/issues/56, I tried iterating over every polygon component. The API fails either because there is no source data or the intersect does not provide any matches (I tried both centroid and touches). Can we investigate which is the case? Either way, it should raise an exception explaining the issue. Right now it still returns an empty array.

**To Reproduce**

```
from typing import Dict
import requests
import geopandas as gpd
import pandas as pd
from geojson_pydantic import Feature, Polygon
import json
from shapely.geometry import shape

BASE_URL = "https://space2stats.ds.io"
FIELDS_ENDPOINT = f"{BASE_URL}/fields"
SUMMARY_ENDPOINT = f"{BASE_URL}/summary"

iso3 = 'KEN'
adm = 'ADM1'
url = f'https://www.geoboundaries.org/api/current/gbOpen/{iso3}/{adm}/'
res = requests.get(url).json()
adm = gpd.read_file(res['gjDownloadURL'])
adm = adm.loc[[42]].copy()
adm = adm.explode(index_parts=False)
adm.reset_index(inplace=True, drop=True)

AOIModel = Feature[Polygon, Dict]

geojson_str = adm.to_json()
adm_geojson = json.loads(geojson_str)
adm_features = adm_geojson['features']

gdfs = []
for feature in adm_features:

    # I think this validates the geojson
    feat = AOIModel(**feature)
    
    # Define the Request Payload
    request_payload = {
        "aoi": feature,
        "spatial_join_method": "touches",
        "fields": ["sum_pop_2020"], 
        "geometry": "polygon"
    }

    # Get Summary Data
    response = requests.post(SUMMARY_ENDPOINT, json=request_payload)
    if response.status_code != 200:
        raise Exception(f"Failed to get summary: {response.text}")
        
    summary_data = response.json()
    if summary_data == []: # Error needs to be added to API
        print(f"Failed to get summary for {feature['id']}")
        continue

    df = pd.DataFrame(summary_data)
    df['adm_id'] = int(feature['id'])
    df['adm_name'] = feature['properties']['shapeName']
    df['geometry'] = df['geometry'].apply(lambda geom: shape(geom))
    gdf = gpd.GeoDataFrame(df, geometry='geometry', crs='EPSG:4326')
    gdfs.append(gdf)

```

### PR DESCRIPTION
**Pull request by zacharyDez**
## What I Changed

- Implemented the `_find_touches` function to identify H3 hexagon IDs that intersect with the boundary of the polygon defined by the AOI GeoJSON and include neighboring hexagons.
- Refactored the `_recursive_polyfill` method to handle cases where no H3 IDs are returned, allowing it to continue to higher resolutions until a valid result is found.
- Enhanced the logic to ensure the correct filtering of H3 hexagon IDs based on spatial join methods: `within` and `touches`.

## How to Test It

- See `test_h3_utils.py`
- Use the `generate_h3_ids` function with various Polygon GeoJSON inputs to ensure that H3 IDs are generated correctly for different geographic shapes.
- Verify that the filtering based on spatial join methods works as expected, particularly with the new `_find_touches` function.
- Run the existing unit tests to confirm that no regressions have occurred in the functionality.

Quick example running with the `touches` keyword:
image

## Other Notes

- Closes #58
- I somehow started from a branch that didn't have the support for multipolygon integrated. I'll resolve the merge conflict and see how to add multipolygon support tomorrow. cc: @alukach

### CODE REVIEW

**Review by vincentsarago — APPROVED** [review]

File: space2stats_api/src/space2stats/h3_utils.py [hunk 1]
```diff
@@ -1,67 +1,122 @@
...
+MAX_RESOLUTION = 15
+
+
+def _get_parent(
+    h3_ids: Iterable[str], resolution: int, target_resolution: int
+) -> Set[str]:
+    if resolution == target_resolution:
+        return set(h3_ids)
+
+    assert resolution > target_resolution
```
Comment (vincentsarago) [inline space2stats_api/src/space2stats/h3_utils.py:16]: maybe add an error message here 

```python
assert resolution > target_resolution, "error message that tells why"
```

File: space2stats_api/src/space2stats/h3_utils.py [hunk 2]
```diff
@@ -1,67 +1,122 @@
...
+            if h3.h3_indexes_are_neighbors(h3_id, n_h3_id) and aoi_shape.intersects(
+                boundary
+            ):
+                neighbors.add(n_h3_id)
+
+    return neighbors.union(h3_ids)
+
+
+def _generate_h3_ids(
+    aoi_geojson: Dict[str, Any], resolution: int, spatial_join_method: str
```
Comment (vincentsarago) [inline space2stats_api/src/space2stats/h3_utils.py:71]: we can define `spatial_join_method` with type `Literal["touche", "within", "centroid"]`

File: space2stats_api/src/space2stats/h3_utils.py [hunk 3]
```diff
@@ -1,67 +1,122 @@
...
+) -> Set[str]:
+    if spatial_join_method not in ["touches", "within", "centroid"]:
+        raise ValueError("Invalid spatial join method")
```
Comment (vincentsarago) [inline space2stats_api/src/space2stats/h3_utils.py:74]:
```suggestion
        raise ValueError(f"Invalid spatial join method: {spatial_join_method}")
```

File: space2stats_api/src/space2stats/h3_utils.py (hunk 1 above)
Comment (zacharyDez) [reply space2stats_api/src/space2stats/h3_utils.py:16]: We can actually remove the `_get_parent` function!

File: space2stats_api/src/space2stats/h3_utils.py (hunk 2 above)
Comment (zacharyDez) [reply space2stats_api/src/space2stats/h3_utils.py:71]: great suggestion!

File: space2stats_api/src/space2stats/h3_utils.py (hunk 3 above)
Comment (zacharyDez) [reply space2stats_api/src/space2stats/h3_utils.py:74]: I like it!

**Review by alukach — APPROVED** [review]

**Pull request merged**